File: examples/java/src/c/Persistent.java
package c;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.ArrayList;
import java.util.stream.Collectors;
public class Persistent {
public static final long testMap() {
HashMap<String,Long> tmp1 = new HashMap<String,Long>();
tmp1.put(":x", 11L);
tmp1.put(":y", 13L);
final HashMap<String,Long> a = tmp1;
HashMap<Object,Object> tmp2 = new HashMap<Object,Object>();
tmp2.put(":x", 11L);
tmp2.put(":y", 13L);
final HashMap<Object,Object> b = tmp2;
final io.lacuna.bifurcan.Map<String,Long> c = new io.lacuna.bifurcan.Map<String,Long>().put(":x", 11L, io.lacuna.bifurcan.Maps.MERGE_LAST_WRITE_WINS).put(":y", 13L, io.lacuna.bifurcan.Maps.MERGE_LAST_WRITE_WINS);
final io.lacuna.bifurcan.Map<Object,Object> d = new io.lacuna.bifurcan.Map<Object,Object>().put(":x", 11L, io.lacuna.bifurcan.Maps.MERGE_LAST_WRITE_WINS).put(":y", 13L, io.lacuna.bifurcan.Maps.MERGE_LAST_WRITE_WINS);
final Object e = {:y 13, :x 11};
{
System.out.println(("" + "key :y in mutable map a returns " + a.get(":y")));
{
final Object anyY = ":y";
final long getBAnyY = (long)b.get(anyY);
System.out.println(("" + "key :y in mutable map b returns " + getBAnyY));
}
System.out.println(("" + "key :y in persistent map c returns " + c.get(":y", null)));
{
final Object anyY = ":y";
final long getDAnyY = (long)d.get(anyY, null);
System.out.println(("" + "key :y in persistent map d returns " + getDAnyY));
}
{
final Object anyY = ":y";
final io.lacuna.bifurcan.Map<Object,Object> eMap = (io.lacuna.bifurcan.Map<Object,Object>)e;
final long getEAnyY = (long)eMap.get(anyY, null);
System.out.println(("" + "key :y in persistent map e returns " + getEAnyY));
}
return 3L;
}
}
public static final long testVector() {
ArrayList<Long> tmp3 = new ArrayList<Long>();
tmp3.add(11L);
tmp3.add(13L);
final ArrayList<Long> a = tmp3;
ArrayList<Object> tmp4 = new ArrayList<Object>();
tmp4.add(11L);
tmp4.add(13L);
final ArrayList<Object> b = tmp4;
final io.lacuna.bifurcan.List<Long> c = new io.lacuna.bifurcan.List<Long>().addLast(11L).addLast(13L);
final io.lacuna.bifurcan.List<Object> d = new io.lacuna.bifurcan.List<Object>().addLast(11L).addLast(13L);
{
System.out.println(("" + "size of mutable vector a returns " + a.size()));
System.out.println(("" + "size of mutable vector b returns " + b.size()));
System.out.println(("" + "size of persistent vector c returns " + c.size()));
System.out.println(("" + "size of persistent vector d returns " + d.size()));
return 5L;
}
}
public static final long testSet() {
HashSet<Long> tmp5 = new HashSet<Long>();
tmp5.add(11L);
tmp5.add(13L);
tmp5.add(15L);
final HashSet<Long> a = tmp5;
HashSet<Object> tmp6 = new HashSet<Object>();
tmp6.add(11L);
tmp6.add(13L);
tmp6.add(15L);
final HashSet<Object> b = tmp6;
final io.lacuna.bifurcan.Set<Long> c = new io.lacuna.bifurcan.Set<Long>().add(11L).add(13L).add(15L);
final io.lacuna.bifurcan.Set<Object> d = new io.lacuna.bifurcan.Set<Object>().add(11L).add(13L).add(15L);
{
System.out.println(("" + "size of mutable set a returns " + a.size()));
System.out.println(("" + "size of mutable set b returns " + b.size()));
System.out.println(("" + "size of persistent set c returns " + c.size()));
System.out.println(("" + "size of persistent set d returns " + d.size()));
return 7L;
}
}
public static final void main(String[] args) {
System.out.println(c.Persistent.testMap());
System.out.println(c.Persistent.testVector());
System.out.println(c.Persistent.testSet());
}
}

File: examples/rust/src/c/persistent.rs
use crate::kalai::kalai;
use crate::kalai::kalai::PMap;
pub fn test_map() -> i64 {
    let a: std::collections::HashMap<String, i64> = {
        let mut tmp1: std::collections::HashMap<String, i64> = std::collections::HashMap::new();
        tmp1.insert(String::from(":x"), 11i64);
        tmp1.insert(String::from(":y"), 13i64);
        tmp1
    };
    let b: std::collections::HashMap<kalai::BValue, kalai::BValue> = {
        let mut tmp2: std::collections::HashMap<kalai::BValue, kalai::BValue> =
            std::collections::HashMap::new();
        tmp2.insert(
            kalai::BValue::from(String::from(":x")),
            kalai::BValue::from(11i64),
        );
        tmp2.insert(
            kalai::BValue::from(String::from(":y")),
            kalai::BValue::from(13i64),
        );
        tmp2
    };
    let c: rpds::HashTrieMap<String, i64> = rpds::HashTrieMap::new()
        .insert(String::from(":x"), 11i64)
        .insert(String::from(":y"), 13i64);
    let d: rpds::HashTrieMap<kalai::BValue, kalai::BValue> = rpds::HashTrieMap::new()
        .insert(
            kalai::BValue::from(String::from(":x")),
            kalai::BValue::from(11i64),
        )
        .insert(
            kalai::BValue::from(String::from(":y")),
            kalai::BValue::from(13i64),
        );
    let e: kalai::BValue = kalai::BValue::from(
        rpds::HashTrieMap::new()
            .insert(
                kalai::BValue::from(String::from(":x")),
                kalai::BValue::from(11i64),
            )
            .insert(
                kalai::BValue::from(String::from(":y")),
                kalai::BValue::from(13i64),
            ),
    );
    {
        println!(
            "{}",
            format!(
                "{}{}",
                String::from("key :y in mutable map a returns "),
                a.get(&String::from(":y")).unwrap().clone()
            )
        );
        {
            let any_y: kalai::BValue = kalai::BValue::from(String::from(":y"));
            let get_b_any_y: i64 = i64::from(b.get(&any_y).unwrap().clone());
            println!(
                "{}",
                format!(
                    "{}{}",
                    String::from("key :y in mutable map b returns "),
                    get_b_any_y
                )
            );
        }
        println!(
            "{}",
            format!(
                "{}{}",
                String::from("key :y in persistent map c returns "),
                c.get(&String::from(":y")).unwrap().clone()
            )
        );
        {
            let any_y: kalai::BValue = kalai::BValue::from(String::from(":y"));
            let get_d_any_y: i64 = i64::from(d.get(&any_y).unwrap().clone());
            println!(
                "{}",
                format!(
                    "{}{}",
                    String::from("key :y in persistent map d returns "),
                    get_d_any_y
                )
            );
        }
        {
            let any_y: kalai::BValue = kalai::BValue::from(String::from(":y"));
            let e_map: rpds::HashTrieMap<kalai::BValue, kalai::BValue> = rpds::HashTrieMap::from(e);
            let get_e_any_y: i64 = i64::from(e_map.get(&any_y).unwrap().clone());
            println!(
                "{}",
                format!(
                    "{}{}",
                    String::from("key :y in persistent map e returns "),
                    get_e_any_y
                )
            );
        }
        return 3i64;
    }
}
pub fn test_vector() -> i64 {
    let a: std::vec::Vec<i64> = {
        let mut tmp3: std::vec::Vec<i64> = std::vec::Vec::new();
        tmp3.push(11i64);
        tmp3.push(13i64);
        tmp3
    };
    let b: std::vec::Vec<kalai::BValue> = {
        let mut tmp4: std::vec::Vec<kalai::BValue> = std::vec::Vec::new();
        tmp4.push(kalai::BValue::from(11i64));
        tmp4.push(kalai::BValue::from(13i64));
        tmp4
    };
    let c: rpds::Vector<i64> = rpds::Vector::new().push_back(11i64).push_back(13i64);
    let d: rpds::Vector<kalai::BValue> = rpds::Vector::new()
        .push_back(kalai::BValue::from(11i64))
        .push_back(kalai::BValue::from(13i64));
    {
        println!(
            "{}",
            format!(
                "{}{}",
                String::from("size of mutable vector a returns "),
                a.len() as i32
            )
        );
        println!(
            "{}",
            format!(
                "{}{}",
                String::from("size of mutable vector b returns "),
                b.len() as i32
            )
        );
        println!(
            "{}",
            format!(
                "{}{}",
                String::from("size of persistent vector c returns "),
                c.len() as i32
            )
        );
        println!(
            "{}",
            format!(
                "{}{}",
                String::from("size of persistent vector d returns "),
                d.len() as i32
            )
        );
        return 5i64;
    }
}
pub fn test_set() -> i64 {
    let a: std::collections::HashSet<i64> = {
        let mut tmp5: std::collections::HashSet<i64> = std::collections::HashSet::new();
        tmp5.insert(11i64);
        tmp5.insert(13i64);
        tmp5.insert(15i64);
        tmp5
    };
    let b: std::collections::HashSet<kalai::BValue> = {
        let mut tmp6: std::collections::HashSet<kalai::BValue> = std::collections::HashSet::new();
        tmp6.insert(kalai::BValue::from(11i64));
        tmp6.insert(kalai::BValue::from(13i64));
        tmp6.insert(kalai::BValue::from(15i64));
        tmp6
    };
    let c: rpds::HashTrieSet<i64> = rpds::HashTrieSet::new()
        .insert(11i64)
        .insert(13i64)
        .insert(15i64);
    let d: rpds::HashTrieSet<kalai::BValue> = rpds::HashTrieSet::new()
        .insert(kalai::BValue::from(11i64))
        .insert(kalai::BValue::from(13i64))
        .insert(kalai::BValue::from(15i64));
    {
        println!(
            "{}",
            format!(
                "{}{}",
                String::from("size of mutable set a returns "),
                a.len() as i32
            )
        );
        println!(
            "{}",
            format!(
                "{}{}",
                String::from("size of mutable set b returns "),
                b.len() as i32
            )
        );
        println!(
            "{}",
            format!(
                "{}{}",
                String::from("size of persistent set c returns "),
                c.size() as i32
            )
        );
        println!(
            "{}",
            format!(
                "{}{}",
                String::from("size of persistent set d returns "),
                d.size() as i32
            )
        );
        return 7i64;
    }
}
pub fn main() {
    let _args: std::vec::Vec<String> = std::env::args().collect();
    {
        println!("{}", test_map());
        println!("{}", test_vector());
        println!("{}", test_set());
    }
}
